### PR TITLE
core: add functional-type directive

### PR DESCRIPTION
**Pull request by alexarice**
Adds the functional-type directive, which takes two typeable directives as input.

The key difficulty here was dealing with the results, which are usually surrounded by parentheses but don't have to be if there is exactly one result.

Still a draft as this needs more tests, also stacked on #3516.

### CODE REVIEW

**Review by superlopuh — COMMENTED** [review]
Very very nice

**Comment by alexarice** [comment]
I've added an extra test to this. This also has given a way to actually trigger `parse_single_type` on the `operands` and `results` directives, which has revealed that they weren't quite implemented correctly. They now use the same assignment function as their many valued counterparts to be more robust.

File: tests/irdl/test_declarative_assembly_format.py [hunk 1]
```diff
@@ -1004,6 +1004,35 @@ class TwoOperandsOp(IRDLOperation):
...
 
 
+@pytest.mark.parametrize(
+    "program",
+    [
+        "%0 = test.functional_type %1, %2 : (i32, i32) -> i32",
+        "test.functional_type %0, %1 : (i32, i32) -> ()",
+        "%0, %1 = test.functional_type %2, %3 : (i32, i32) -> (i32, i32)",
+        "%0 = test.functional_type %1 : (i32) -> i32",
+        "%0 = test.functional_type  : () -> i32",
```
[the remark this answers is not in the document]
Comment (alexarice) [reply tests/irdl/test_declarative_assembly_format.py]: Should be fixed by #3530

File: xdsl/irdl/declarative_assembly_format.py [hunk 2]
```diff
@@ -660,9 +660,11 @@ def parse_optional(self, parser: Parser, state: ParsingState) -> bool:
...
 
     def parse_single_type(self, parser: Parser, state: ParsingState) -> None:
-        if len(state.operand_types) > 1:
-            parser.raise_error("Expected multiple types but received one.")
-        state.operand_types[0] = parser.parse_type()
+        pos_start = parser.pos
+        if s := self._set_using_variadic_index(
+            state.operand_types, "operand types", (parser.parse_type(),)
+        ):
+            parser.raise_error(s, at_position=pos_start, end_position=parser.pos)
```
Comment (superlopuh) [inline xdsl/irdl/declarative_assembly_format.py:667]: should these then be in a separate PR? With a separate test to demonstrate the impact of the change?

Comment (alexarice) [reply xdsl/irdl/declarative_assembly_format.py:667]: Maybe, though I'm not quite sure how to test them, as the `functional-type` directive is the only thing that actually uses these functions

Comment (alexarice) [reply xdsl/irdl/declarative_assembly_format.py:667]: I have an idea, will cook something up

**Comment by alexarice** [comment]
I've stacked this on #3553 , separating out the changes mentioned in the comment above

File: tests/irdl/test_declarative_assembly_format.py [hunk 3]
```diff
@@ -1758,6 +1758,66 @@ def print(self, printer: Printer):
...
+        name = "test.functional_type"
+
+        ops = var_operand_def()
+        res = var_result_def()
+
+        assembly_format = "$ops attr-dict `:` functional-type($ops, $res)"
+
+    ctx = MLContext()
+    ctx.load_op(FunctionalTypeOp)
+
```
[the remark this answers is not in the document]
Comment (alexarice) [reply tests/irdl/test_declarative_assembly_format.py:1790]: Fixed

File: xdsl/irdl/declarative_assembly_format.py (hunk 2 above)
Comment (alexarice) [reply xdsl/irdl/declarative_assembly_format.py]: Now separated

File: tests/irdl/test_declarative_assembly_format.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (alexarice) [reply tests/irdl/test_declarative_assembly_format.py]: Is now fixed

**Review by superlopuh — APPROVED** [review]
Beautiful

**Review by kimxworrall — COMMENTED** [review]
Nice

**Review by kimxworrall — APPROVED** [review]
Nice :)
